File: shared/time/time.go
// Package time is a dependency-injected time source modeled on TigerBeetle's
// vsr.Time, whose backend is a struct of function pointers (a vtable). Here that is
// a struct of closures: production wires an OS clock (time/default), a simulation
// wires a Virtual one, and the code between never knows which it holds. The Virtual
// backend lives here because it is pure arithmetic with no operating-system call.
package time

import (
	invariant "local/james-orcales/shared/invariant/default"
	"local/james-orcales/shared/math/fixedpoint"
)

// INTEGER_64_MINIMUM is the smallest signed 64-bit integer. A Moment, a Duration, and
// a Tick_Count each occupy the whole signed 64-bit range: a virtual clock reads its
// epoch, its skew, and its tick count as the caller sets them, and none of the three
// carries a narrower bound than its storage.
const INTEGER_64_MINIMUM int64 = -9223372036854775808

// INTEGER_64_MAXIMUM is the largest signed 64-bit integer.
const INTEGER_64_MAXIMUM int64 = 9223372036854775807

// Moment is a clock reading in nanoseconds since an arbitrary, clock-specific epoch
// (TigerBeetle's stdx.Instant). Only the difference between two Moments from the
// SAME clock is meaningful; a monotonic Moment and a realtime Moment are not
// comparable.
type Moment int64

// Moment_Invariants states the complete clock-reading domain.
func Moment_Invariants(moment Moment, namespace invariant.Namespace) {
	invariant.Tree(moment, namespace).
		Range_Int64(int64(moment), INTEGER_64_MINIMUM, INTEGER_64_MAXIMUM).
		Ensure()
}

// Duration is a span of nanoseconds (TigerBeetle's stdx.Duration).
type Duration int64

// Duration_Invariants states the complete nanosecond-span domain.
func Duration_Invariants(duration Duration, namespace invariant.Namespace) {
	invariant.Tree(duration, namespace).
		Range_Int64(int64(duration), INTEGER_64_MINIMUM, INTEGER_64_MAXIMUM).
		Ensure()
}

// NANOSECOND is the unit a Duration counts in.
const NANOSECOND Duration = 1

// MICROSECOND is a thousand nanoseconds.
const MICROSECOND = NANOSECOND * 1000

// MILLISECOND is a thousand microseconds.
const MILLISECOND = MICROSECOND * 1000

// SECOND is a thousand milliseconds.
const SECOND = MILLISECOND * 1000

// MINUTE is sixty seconds.
const MINUTE = SECOND * 60

// HOUR is sixty minutes.
const HOUR = MINUTE * 60

// DAY is twenty-four hours.
const DAY = HOUR * 24

// WEEK is seven days.
const WEEK = DAY * 7

// Clock is the injected time source — the Go translation of TigerBeetle's `Time`
// vtable, expressed as closures so the backend is chosen by value. It is read-only:
// advancing time is the driver's job (the tick returned beside the clock at
// construction), so a holder can only read the current Moment, never move time.
type Clock struct {
	// Now_Monotonic reads the monotonic clock, which never regresses; use it to
	// measure elapsed time, timeouts, and latency.
	Now_Monotonic func() (moment Moment)
	// Now_Realtime reads wall-clock time as nanoseconds since the Unix epoch; it can
	// jump, so use it only for calendar timestamps, never for elapsed time.
	Now_Realtime func() (moment Moment)
}

// Clock_Invariants states that both readers are bound. A Clock is a vtable, so its
// only property is that every slot is filled: the zero Clock reads as a Clock but
// panics on first use, and a backend that fills one slot and forgets the other is the
// same failure one call later.
func Clock_Invariants(clock Clock, namespace invariant.Namespace) {
	invariant.Always(
		clock.Now_Monotonic != nil, "A Clock has a monotonic reader.",
	)
	invariant.Always(
		clock.Now_Realtime != nil, "A Clock has a realtime reader.",
	)
}

// Tick_Count is how many times a virtual clock advanced — the abscissa every skew
// model reads (TimeSim's x).
type Tick_Count int64

// Tick_Count_Invariants states the complete tick-count domain.
func Tick_Count_Invariants(ticks Tick_Count, namespace invariant.Namespace) {
	invariant.Tree(ticks, namespace).
		Range_Int64(int64(ticks), INTEGER_64_MINIMUM, INTEGER_64_MAXIMUM).
		Ensure()
}

// Offset models how a simulated wall clock deviates from true elapsed time —
// TigerBeetle's TimeSim.offset. It is what makes Now_Realtime diverge from
// Now_Monotonic. A nil Offset is a perfect clock.
type Offset func(ticks Tick_Count) (skew Duration)

// Virtual_Clock configures the deterministic clock Virtual_Clock_To_Clock builds —
// TigerBeetle's TimeSim. Time advances only when Tick is called, so a simulation
// reaches a future Moment by ticking rather than by waiting.
type Virtual_Clock struct {
	// Resolution is how far the monotonic clock advances on each Tick — the grain of
	// a simulated oscillator (TimeSim.resolution).
	Resolution Duration
	// Epoch is the wall-clock origin: Now_Realtime at tick zero, before any skew
	// (TimeSim.epoch).
	Epoch Moment
	// Skew bends Now_Realtime away from true elapsed time; nil is a perfect clock
	// (TimeSim.offset).
	Skew Offset
}

// Virtual_Clock_Invariants states the two scalars a virtual clock is configured with.
// Skew is a closure, so the arithmetic it stands for has no domain to state here; its
// coefficients are stated where Skew builds it.
func Virtual_Clock_Invariants(virtual Virtual_Clock, namespace invariant.Namespace) {
	Duration_Invariants(virtual.Resolution, namespace)
	Moment_Invariants(virtual.Epoch, namespace)
}

// Virtual_Clock_To_Clock returns a read-only Clock backed by a deterministic, OS-free
// virtual clock, plus the tick that advances it. The clock's closures and tick share
// one counter, so tick advances what the next Now_Monotonic reads. Only the driver —
// package main or a test harness — holds tick; pure code holds only the Clock and so
// can read time but never move it.
func Virtual_Clock_To_Clock(virtual Virtual_Clock) (clock Clock, tick func()) {
	defer func() { Clock_Invariants(clock, "virtual_clock_to_clock.clock") }()
	Virtual_Clock_Invariants(virtual, "virtual_clock_to_clock.virtual")
	ticks := Tick_Count(0)
	clock = Clock{
		Now_Monotonic: func() (moment Moment) {
			return Moment(int64(ticks) * int64(virtual.Resolution))
		},
		Now_Realtime: func() (moment Moment) {
			now := virtual.Epoch + Moment(int64(ticks)*int64(virtual.Resolution))
			if virtual.Skew == nil {
				return now
			}
			return now - Moment(virtual.Skew(ticks))
		},
	}
	return clock, func() { ticks++ }
}

// SKEW_KIND_LINEAR models constant drift: A nanoseconds of skew per tick plus an
// initial B (TimeSim OffsetType.linear, A*x + B).
const SKEW_KIND_LINEAR Skew_Kind = 0

// SKEW_KIND_PERIODIC models a sinusoidal wobble of amplitude A over a period of B
// ticks (TimeSim OffsetType.periodic, A*sin(x*2pi/B)).
const SKEW_KIND_PERIODIC Skew_Kind = 1

// SKEW_KIND_STEP models a discontinuous jump of A after B ticks — an NTP correction
// or operator clock change (TimeSim OffsetType.step).
const SKEW_KIND_STEP Skew_Kind = 2

// Skew_Kind selects which clock-deviation model Skew builds.
type Skew_Kind uint8

// Skew_Kind_Invariants holds a kind to the three models Skew builds. The default arm
// of that switch is the linear model, so an unlisted kind would drift silently rather
// than fail.
func Skew_Kind_Invariants(kind Skew_Kind, namespace invariant.Namespace) {
	invariant.Tree(kind, namespace).
		Enum_3_Uint8(
			uint8(kind),
			uint8(SKEW_KIND_LINEAR),
			uint8(SKEW_KIND_PERIODIC),
			uint8(SKEW_KIND_STEP),
		).
		Ensure()
}

// Skew_Input is the model and its coefficients, mirroring TimeSim's offset_type plus
// offset_coefficient_A and offset_coefficient_B.
type Skew_Input struct {
	// Kind selects the deviation model.
	Kind Skew_Kind
	// A is the magnitude coefficient: drift-per-tick, amplitude, or step size.
	A Duration
	// B is the tick coefficient: the linear initial offset, the periodic period, or
	// the step's onset tick.
	B Tick_Count
}

// Skew_Input_Invariants states the model and each of its two coefficients.
func Skew_Input_Invariants(input Skew_Input, namespace invariant.Namespace) {
	Skew_Kind_Invariants(input.Kind, namespace)
	Duration_Invariants(input.A, namespace)
	Tick_Count_Invariants(input.B, namespace)
}

// Skew builds the Offset described by input.
func Skew(input Skew_Input) (offset Offset) {
	Skew_Input_Invariants(input, "skew.input")
	switch input.Kind {
	case SKEW_KIND_PERIODIC:
		return func(ticks Tick_Count) (skew Duration) {
			// A zero period is a degenerate sinusoid; report no skew rather than divide
			// (or take a remainder) by zero.
			if input.B == 0 {
				return 0
			}
			// Reduce the phase to one period before lifting it into fixed-point, so a
			// long-running tick count cannot overflow the scaled numerator.
			phase := ticks % input.B
			turns := fixedpoint.From_Ratio(
				fixedpoint.Numerator(phase),
				fixedpoint.Denominator(input.B),
			)
			amplitude := fixedpoint.Number(fixedpoint.From_Integer(
				fixedpoint.Whole_Integer(input.A),
			))
			wobble := fixedpoint.Multiply(
				fixedpoint.Multiplicand(amplitude),
				fixedpoint.Multiplier(fixedpoint.Sine_Turns(turns)),
			)
			return Duration(fixedpoint.Whole(wobble))
		}
	case SKEW_KIND_STEP:
		return func(ticks Tick_Count) (skew Duration) {
			if ticks > input.B {
				return input.A
			}
			return 0
		}
	default:
		return func(ticks Tick_Count) (skew Duration) {
			return Duration(ticks)*input.A + Duration(input.B)
		}
	}
}

File: shared_rs/src/time.rs
//! A dependency-injected time source modeled on TigerBeetle's `vsr.Time`. One
//! [`Clock`] is injected; calling code reads it with [`now_monotonic`] /
//! [`now_realtime`] and advances it with [`tick`] / [`sleep`] and never knows
//! whether it holds the deterministic virtual backend (a simulation) or the host
//! operating system (production). The Go original is a vtable of closures sharing
//! a mutated tick counter; the dialect forbids that shared mutation, so the
//! backend is an enum and advancing is consume-and-return — `tick` takes a
//! `Clock` and returns the advanced one — rather than a closure mutating in place.
//!
//! ```
//! use shared_rs::time;
//! let clock = time::new_virtual(time::SECOND, time::Moment(0), time::Skew::Uniform);
//! assert_eq!(time::now_monotonic(&clock), time::Moment(0));
//! let clock = time::tick(clock);
//! assert_eq!(time::now_monotonic(&clock), time::Moment(time::SECOND.0));
//! ```

use crate::fixedpoint;
use std::thread;
use std::time as wallclock;

/// A clock reading in nanoseconds since an arbitrary, clock-specific epoch. Only
/// the difference between two `Moment`s from the SAME clock is meaningful; a
/// monotonic and a realtime `Moment` are not comparable.
#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
pub struct Moment(pub i64);

/// A span of nanoseconds.
#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
pub struct Duration(pub i64);

/// The unit a [`Duration`] counts in.
pub const NANOSECOND: Duration = Duration(1);
/// A thousand nanoseconds.
pub const MICROSECOND: Duration = Duration(NANOSECOND.0 * 1000);
/// A thousand microseconds.
pub const MILLISECOND: Duration = Duration(MICROSECOND.0 * 1000);
/// A thousand milliseconds.
pub const SECOND: Duration = Duration(MILLISECOND.0 * 1000);
/// Sixty seconds.
pub const MINUTE: Duration = Duration(SECOND.0 * 60);
/// Sixty minutes.
pub const HOUR: Duration = Duration(MINUTE.0 * 60);
/// Twenty-four hours.
pub const DAY: Duration = Duration(HOUR.0 * 24);
/// Seven days.
pub const WEEK: Duration = Duration(DAY.0 * 7);

/// The `Moment` a span past another.
pub fn add(moment: Moment, span: Duration) -> Moment {
    Moment(moment.0 + span.0)
}

/// The span from `earlier` to `later`. Meaningful only when both come from the
/// same clock.
pub fn since(later: Moment, earlier: Moment) -> Duration {
    Duration(later.0 - earlier.0)
}

/// The injected time source. Both backends share one read/advance API, so calling
/// code is written against `Clock` and never branches on which it holds — the Go
/// `Clock` vtable, expressed as an enum the dialect can build without the mutable
/// closures the original used.
#[derive(Copy, Clone, Debug)]
pub enum Clock {
    /// The deterministic, system-free backend: time advances only on `tick`/`sleep`.
    Virtual(Virtual_Clock),
    /// The operating-system backend, holding the monotonic reference point captured
    /// at creation (Rust's `Instant` exposes only elapsed time, not an absolute).
    Operating_System(wallclock::Instant),
}

/// A fresh virtual clock at tick zero, wrapped as a [`Clock`]. Pass `Skew::Uniform`
/// for a perfect clock.
pub fn new_virtual(resolution: Duration, epoch: Moment, skew: Skew) -> Clock {
    Clock::Virtual(Virtual_Clock { resolution, epoch, skew, ticks: 0 })
}

/// A clock backed by the host operating system, wrapped as a [`Clock`].
pub fn new_operating_system() -> Clock {
    Clock::Operating_System(wallclock::Instant::now())
}

/// The monotonic reading — never regresses, so use it for elapsed time, timeouts,
/// and latency. The virtual backend reads tick count times resolution; the
/// operating-system backend reads nanoseconds elapsed since the reference point.
pub fn now_monotonic(clock: &Clock) -> Moment {
    match clock {
        Clock::Virtual(virtual_clock) => {
            Moment(virtual_clock.ticks * virtual_clock.resolution.0)
        }
        Clock::Operating_System(epoch) => Moment(epoch.elapsed().as_nanos() as i64),
    }
}

/// The wall-clock reading — it can jump, so use it only for calendar timestamps,
/// never for elapsed time. The virtual backend is epoch plus elapsed less any
/// modeled skew; the operating-system backend reads nanoseconds since the Unix
/// epoch.
pub fn now_realtime(clock: &Clock) -> Moment {
    match clock {
        Clock::Virtual(virtual_clock) => virtual_realtime(virtual_clock),
        Clock::Operating_System(_) => operating_system_realtime(),
    }
}

/// Advances the clock by one resolution, returning the advanced clock. Ticking the
/// operating-system backend is a no-op — it advances on its own.
pub fn tick(clock: Clock) -> Clock {
    match clock {
        Clock::Virtual(virtual_clock) => {
            Clock::Virtual(Virtual_Clock { ticks: virtual_clock.ticks + 1, ..virtual_clock })
        }
        Clock::Operating_System(epoch) => Clock::Operating_System(epoch),
    }
}

/// Waits out the span and returns the clock. The virtual backend advances its own
/// time instead of waiting (a simulation never blocks); the operating-system
/// backend blocks the thread for the real duration.
pub fn sleep(clock: Clock, duration: Duration) -> Clock {
    match clock {
        Clock::Virtual(virtual_clock) => Clock::Virtual(Virtual_Clock {
            ticks: virtual_clock.ticks + duration.0 / virtual_clock.resolution.0,
            ..virtual_clock
        }),
        Clock::Operating_System(epoch) => {
            operating_system_sleep(duration);
            Clock::Operating_System(epoch)
        }
    }
}

/// The deterministic backend of a [`Clock`]: time advances only when [`tick`] or
/// [`sleep`] is called, so a simulation reaches a future `Moment` by ticking
/// rather than by waiting on a wall clock.
#[derive(Copy, Clone, Debug)]
pub struct Virtual_Clock {
    /// How far the monotonic clock advances on each tick — the grain of a
    /// simulated oscillator.
    pub resolution: Duration,
    /// The wall-clock origin: realtime at tick zero, before any skew.
    pub epoch: Moment,
    /// How realtime bends away from true elapsed time; `Skew::Uniform` is a perfect
    /// clock.
    pub skew: Skew,
    /// The current tick count; reads are a pure function of it.
    pub ticks: i64,
}

fn virtual_realtime(clock: &Virtual_Clock) -> Moment {
    // `Skew::Uniform` contributes zero, so realtime is just epoch plus elapsed.
    let now = Moment(clock.epoch.0 + clock.ticks * clock.resolution.0);
    Moment(now.0 - skew_at(&clock.skew, clock.ticks).0)
}

fn operating_system_realtime() -> Moment {
    let since_epoch = wallclock::SystemTime::now()
        .duration_since(wallclock::SystemTime::UNIX_EPOCH)
        .unwrap_or(wallclock::Duration::ZERO);
    Moment(since_epoch.as_nanos() as i64)
}

fn operating_system_sleep(duration: Duration) {
    // A non-positive span is not a wait; only a real one blocks the thread.
    if duration.0 > 0 {
        thread::sleep(wallclock::Duration::from_nanos(duration.0 as u64));
    }
}

/// How a simulated wall clock deviates from true elapsed time. Each model carries
/// its own coefficients, so an invalid pairing cannot be expressed.
#[derive(Copy, Clone, Debug)]
pub enum Skew {
    /// A perfect clock: realtime tracks elapsed time with no deviation.
    Uniform,
    /// Constant drift: `drift_per_tick` nanoseconds each tick, plus a fixed
    /// `initial_offset`.
    Linear { drift_per_tick: Duration, initial_offset: Duration },
    /// A sinusoidal wobble of `amplitude` over a `period` measured in ticks.
    Periodic { amplitude: Duration, period: i64 },
    /// A discontinuous `jump` after `onset_tick` ticks — an NTP correction or
    /// operator clock change.
    Step { jump: Duration, onset_tick: i64 },
}

/// The skew this model produces at the given tick count.
pub fn skew_at(skew: &Skew, ticks: i64) -> Duration {
    match skew {
        Skew::Uniform => Duration(0),
        Skew::Linear { drift_per_tick, initial_offset } => {
            Duration(ticks * drift_per_tick.0 + initial_offset.0)
        }
        Skew::Step { jump, onset_tick } => match ticks > *onset_tick {
            true => *jump,
            false => Duration(0),
        },
        Skew::Periodic { amplitude, period } => periodic_skew(*amplitude, *period, ticks),
    }
}

fn periodic_skew(amplitude: Duration, period: i64, ticks: i64) -> Duration {
    // A zero period is a degenerate sinusoid; report no skew rather than divide by
    // zero. The phase is reduced to one period before the fixed-point lift, so a
    // long tick count cannot overflow the scaled numerator.
    match period == 0 {
        true => Duration(0),
        false => {
            let phase = ticks % period;
            let turns = fixedpoint::divide(
                fixedpoint::from_integer(phase),
                fixedpoint::from_integer(period),
            );
            let wobble = fixedpoint::multiply(
                fixedpoint::from_integer(amplitude.0),
                fixedpoint::sine_turns(turns),
            );
            Duration(fixedpoint::whole(wobble))
        }
    }
}
